### ISSUE TITLE
Scotland missing in "CodemastersDirtRally2.TrackNames.csv"

### ISSUE THREAD

**Issue by Shaftoe62** [SHWotever/SimHub#813]
The "CodemastersDirtRally2.TrackNames.csv" is missing the tracks for Scotland:

12583.41 - South Morningside
5822.77 - Old Butterstone Muir
6967.9 - Rosebank Farm Reverse
12670.59 - South Morningside Reverse
7144.7 - Rosebank Farm
5659.82 - Old Butterstone Muir Reverse
12857.07 - Newhouse Bridge
5245.45 - Glencastle Farm
7587.64 - Annbank Station Reverse
12969.21 - Newhouse Bridge Reverse
7703.72 - Annbank Station
5238.44 - Glencastle Farm Reverse

Jurgen Gardenier

**Comment by SHWotever** [comment]
Hi ! I've reworked the ids in use to get use the full precision, i've been able to match my records and yours to build the full list (except those damn  "Gordolon - Courte Montée" and "Col de Turini - Descente") , 
Here is the list as it will be released for the next update :
[CodemastersDirtRally2.TrackNames.zip](https://github.com/SHWotever/SimHub/files/6489412/CodemastersDirtRally2.TrackNames.zip)

**Comment by Shaftoe62** [comment]
Hi, that looks good.

But does this mean that from now on only the high precision id's will be used to identify tracks?

Because now I'm using the 'old' Id's for most things. Like text files with track info and roadsign-arrays, satellite image names, svg files and so forth.

Most of those I can regenerate using the high precision id's or make an automated rename function or something. 

But you don't know how many users build stuff around the old id's. 

So I don't know...

**Comment by SHWotever** [comment]
Yes indeed it's a breaking change, I try to avoid it as much as possible but from time to time it is required to move forward, at least it will make all Codemasters games as best as possible "future proof". When reviewing the full list it appeared that some totally unrelated stages had really really close values( ie 10805.25 Vs 10805.23).That would be so much easier if the carid/trackid was available but that extra precision should avoid, I hope, any new conflicts. 
I will look to add a toggle to switch between the old I'd mode or the new one. Obviously using the old mode would make you alone to generate the online maps, so it's only a temporary solution.

**Comment by Shaftoe62** [comment]
By the way, about "Gordolon - Courte Montée" and "Col de Turini - Descente":

As their names implicate the first is a climb and the second is a descent. So, even half way through a race the altitude of the first wil be increasing while from the second it will be decreasing.

Perhaps the concepts climb, descent and level (c, d and l) can be incorporated as identifiers.

Of course this won't work for a 100% if you drop in the middle of a race and drive only a few meters, but the more distance you drive the better the odds.

As for map recordings: once a recording is processed at the end of a race you have enough data available to make the distinction. So perhaps you can store a copy of the recording with the added c or d at that moment.

Than you will have these three files availabe:

5175.91015625 (last recording) - unknown
5175.91015625.c (climb) - Gordolon - Courte Montée
5175.91015625.d (decent) - Col de Turini - Descente

SimHub then can then start with the latest recording and switch to climb or descent when enough data has been collected.

Or as soon as SimHub notices that the UDP data for the altitude is completely different from the recording.

Ik know, this seems like a lot of functionality/effort for only two tracks,  but this is an issue that is bound te be repeated in other (codemaster) games in the future.

**Comment by SHWotever** [comment]
Unfortunately it's not something which can be guessed at anytime and instantly. It's really prone to errors in my opinion. The start altitude like you suggested was something more reliable i think (except that it doesn't not cover starting or pausing in the middle). And switching the Id in the middle is not possible too many internal behaviours relies on this id consistency.

**Comment by Shaftoe62** [comment]
I agree that the high precision id's are the way forward.

**Comment by Shaftoe62** [comment]
But as you are implementing the precision id's for tracks and cars in the next version, that would also mean that all the map recordings, including the cloud maps, wil be renamed?

**Comment by SHWotever** [comment]
Yes, indeed, concerning the cloud map I plan to duplicate them before release with the new names so it will be ready out of box, concerning the local map/performance data it will be lost, that's why I plan to add a "backcompatibility" switch in order to stay on the old ids for a smoother transition.

**Comment by Shaftoe62** [comment]
Hi, did you happen to make mapping tables for the old/new track id's, like the one you send me earlier (DirtRally2TrackIDs.xlsx)  for other games as well?

I've been using them for:

Dirt Rally 1
Dirt Rally 2.0
F1-2020

**Comment by SHWotever** [comment]
Hi !
F12017->20 are hard coded based on the game specifications https://forums.codemasters.com/topic/50942-f1-2020-udp-specification/

Dirt rally 1, no I don't have the list

**Comment by Shaftoe62** [comment]
F1: Ah yes, these are simply numbered in "Order of Appearance", forgot.

Dirt Rally 1: Will, after installing the next beta release, both Id's still be available? Or should I best now make a mapping using the old id's and the track length? Assuming the length will be the new id?

**Comment by SHWotever** [comment]
I started the list on DR1 before work :
This is what I have done to date :

Stage | Lenght
-- | --
Pomona Érixi   (Sprint) | 5086.830078125
Ampelonas Ormi (Sprint) | 4860.1904296875
Koryfi Dafni (Sprint) | 4582.009765625
Forkéta Kourva (Sprint) | 4515.39990234375
Paresma Platani (Long) | 10688.08984375
Tsiristra Théa (Long) | 10357.8798828125
Ourea Spevsi (Sprint) | 5739.099609375
Ypsona tou Dasos (Sprint) | 5383.009765625
Abies   Koilàda (Sprint) | 7089.41015625
Pedines   Epidaxi (Sprint) | 6595.31005859375
Anodou Farmakas (Long) | 9666.5
Kathodo Leontiou (Long) | 9665.990234375
Dyffryn   Afon (Sprint) | 5718.099609375
Dyffryn Afon Reverse (Sprint) | 5718.10009765625
Sweet   Lamb (Long) | 9944.8701171875
Geufron Forest (Long) | 10063.58984375
Pant Mawr (Long) | 4788.669921875
Pant Mawr Reverse (Sprint) | 4821.64990234375
Bidno   Moorland (Sprint) | 4993.259765625
Bidno   Moorland Reverse (Sprint) | 5165.9501953125
River   Severn Valley (Long) | 11435.5107421875
Bronfelen   (Long) | 11435.55078125
Fferm   Wynt (Sprint) | 5717.39990234375
Fferm   Wynt Reverse (Sprint) | 5717.3896484375
Col   de Turini Départ (Long) | 9831.970703125
Gordolon   - Courte montée (Sprint) | 5175.91015625
Col de Turini - Descente (Sprint) | 5175.91015625
Col   de Turini - Sprint en montée (Sprint) | 4729.5400390625
Col   de Turini - Sprint en descente (Sprint) | 4730.02001953125
Vallée   descendante (Long) | 10866.8603515625
Route de Turini&nbsp; (Long) | 10805.220703125
Col   de Turini - Départ en descente (Sprint) | 6846.830078125
Route de Turini Montée (Sprint) | 6843.3203125
Route   de Turini Descente (Sprint) | 4015.35986328125
Approche   du Col de Turini - Montée (Sprint) | 3952.15014648438
Pra d'Alart (Long) | 9831.4501953125
Oberstein   (Long) | 11684.169921875
Hammerstein   (Long) | 10805.2392578125

**Comment by Shaftoe62** [comment]
Alright! That's allready a lot.

If you want I can complete the list and share it with you. 

I'm allready working on a unified DR 1&2 id / track info mapping.

That will contain all tracks (including the Rally Cross tracks with their official track names) of both DR 1 and 2 with all track names with their correct accents (Pomona Ékrixi / Östra Hinnsjön / Viñedos dentro del valle Parra, etc).

**Comment by Shaftoe62** [comment]
With that I'm prepared for the next beta update. :-)

**Comment by Shaftoe62** [comment]
Some points of attention:

1. You see two columns of track names. In germany they renamed the "Flugzeugring" to "Innerer Feld-Sprint". So two tracks have different names in 1 & 2.

2. You will notice that some tracks have identical lengths in bot 1 and two while others have different ones. Obviously Codemasters reworked some of the old tracks but copied some others without change.

3. In Dirt Rally 2.0 the Rally Cross tracks no longer have different variants like they had in DR1. At least I cannot find any.

4. I've put the official RX circuit names in a separate column now. I noticed you implemented those as the track names but that will be confusing for users. For example: the track called Mettet in the game is actually on Circuit Jules Tacheny near the village of Mettet in Belgium. But the circuit name is not mentioned in the game.

5. In DR1 you have three main variants of the Pikes Peak Hill Climb stages. 100% Tarmac (modern), Mixed Surface (Tarmac/Gravel) and 100% Gravel (my favorite). The 100% tarmac tracks is different but the other two (Mixed and 100% Gravel) have exactly the same lengths. This is a bit like the "Gordolon conundrum" in Monaco. But here not even xyz coordinates provide a solution because these are identical too. Here I finally had to accept defeat and combine infos for both:

![USA](https://user-images.githubusercontent.com/53898212/119015780-88220d80-b999-11eb-9d46-e1b08bab9aa1.jpg)
